colandr/lib/extractors/locations.py

```diff
@@ -0,0 +1,145 @@
+"""Location extraction from document full text."""
+import logging
+from typing import List
+
+import spacy
+from spacy.tokens import Span
+
+from .metadata import Metadata
+
+
+logger = logging.getLogger(__name__)
+
+
+class LocationExtractor:
+    """Extract locations from document text using spaCy NER."""
+
+    def __init__(self, model: str = "en_core_web_md"):
+        """Initialize the extractor with the specified language model."""
+        self.nlp = spacy.load(model)
+        # Ensure we have sentence segmentation and named entity recognition
+        if not self.nlp.has_pipe("ner"):
+            raise ValueError(f"Model {model} doesn't have NER component")
+
+    def is_in_reference(self, ent: Span) -> bool:
+        """
+        Determine if entity is inside a reference section.
+
+        checks if we're inside parentheses which likely indicates a citation reference.
+        """
+        # Check tokens before this entity for opening parenthesis
+        sent = ent.sent
+        start_idx = ent.start
+
+        for i in range(start_idx - 1, sent.start - 1, -1):
+            if i < 0:
+                break
+            token = sent.doc[i]
+            if token.text == ")":
+                return False
+            if token.text == "(":
+                return True
+
+        return False
+
+    def extract_locations(self, record_id: int, text: str) -> List[Metadata]:
```

Comment:
we're using PY3.11+, so stdlib collections can be used as generics for type annotations -- easier than doing `from typing import List` everywhere!

```suggestion
    def extract_locations(self, record_id: int, text: str) -> list[Metadata]:
```

Reply:
done